### PR TITLE
Disable training with zero num_training_batches when insufficient limit_train_batches

### LINKED ISSUE #5029: Behaviour when limit_train_batches is float is inconsistent
## 🐛 Bug

When passing limit_train_batches as a float to the trainer, the total number of steps during training is inconsistent and is dependent on accumulate_grad_batches, and independent of drop_last. Please see the tests in the notebooks for more examples.

### To Reproduce
See notebook.
https://colab.research.google.com/drive/1RmIwW97NM0bOmF2GUUilPyzCQ2MYYG1g?usp=sharing

### Expected behavior

I don't have an expected behaviour other than to request that the total number of steps remain consistent. Either always drop an incomplete batch, or don't.

### Environment
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     True
	- pyTorch_version:   1.7.0+cu101
	- pytorch-lightning: 1.0.8
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Training should be disabled when `limit_train_batches` is float, and `num_training_batches == int(num_training_batches * limit_train_batches) == 0`, this would be lead to zero `num_training_batches`. But the Training wasn't being skipped, while still doing 1 step per epoch. This PR addresses the issue.

Fixes #5029 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -91,12 +91,8 @@ def num_optimizers(self):
         return num_optimizers
 
     def should_skip_training(self):
-        if self.trainer.current_epoch >= self.trainer.max_epochs:
+        if self.trainer.current_epoch >= self.trainer.max_epochs or self.trainer.num_training_batches == 0:
             return True
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/training_loop.py:95]:
```suggestion
        return self.trainer.current_epoch >= self.trainer.max_epochs or self.trainer.num_training_batches == 0
```

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -91,12 +91,8 @@ def num_optimizers(self):
...
-
-        if self.trainer.limit_train_batches == 0:
-            return True
-
         return False
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/training_loop.py:96]:
```suggestion
```

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py:95]: I think this is enough

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -1444,3 +1445,64 @@ def test_trainer_profiler_incorrect_arg_type(profiler):
                        match=r"Only None, bool, str and subclasses of `BaseProfiler`"
                              r" are valid values for `Trainer`'s `profiler` parameter. *"):
         Trainer(profiler=profiler)
+
+
+@pytest.mark.parametrize(
+    ["limit_train_batches", "global_step", "num_training_batches", "current_epoch", "should_train"],
+    [pytest.param(0.2, 0, 0, 0, False), pytest.param(0.5, 10, 2, 4, True)],
+)
+def test_disabled_training_for_insufficient_limit_train_batches(tmpdir, limit_train_batches, global_step,
```
Comment (rohitgr7) [inline tests/trainer/test_trainer.py:1454]: maybe update `test_disabled_training` in the same file accordingly?

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py:95]: also can you check it `max_steps=0`, does it run?

Comment (kaushikb11) [reply pytorch_lightning/trainer/training_loop.py:95]: Yup, it does run with `max_steps=0`. Will follow up with a fix for that

File: tests/trainer/test_trainer.py (hunk 3 above)
Comment (kaushikb11) [reply tests/trainer/test_trainer.py:1454]: With the similar params? Sure!

Comment (rohitgr7) [reply tests/trainer/test_trainer.py:1454]: yes. you can include one more there with limit_train_batches=0 along with others.

Comment (kaushikb11) [reply tests/trainer/test_trainer.py:1454]: I think it would be better to have it separate as `limit_train_batches=0` would be independent of `dataset_len` and `batch_size`.

Comment (rohitgr7) [reply tests/trainer/test_trainer.py:1454]: ok yeah, no need to include one with `limit_train_batches=0` since it's changed now. you can either remove the old test or update that with this one.

Comment (kaushikb11) [reply tests/trainer/test_trainer.py:1454]: We shouldn't remove it as it tests for `fast_dev_run=True/False` switches.

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -1444,3 +1445,65 @@ def test_trainer_profiler_incorrect_arg_type(profiler):
...
+                                                                num_training_batches, current_epoch, should_train):
+    """
+    Verify when `limit_train_batches` is float & between [0.0, 0.1] and
```
Comment (SeanNaren) [inline tests/trainer/test_trainer.py:1457]: is `[0.0, 0.1]` correct in this case? I assume what we're actually checking for is if there are any training batches to train on in the first place (After accounting for limit_train_batches)

Comment (kaushikb11) [reply tests/trainer/test_trainer.py]: Should be `1.0`, updated. We are testing if we have any training batches to train on with the `limited` training set by `limit_train_batches`. @SeanNaren Is this what you were suggesting?

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -1444,3 +1445,65 @@ def test_trainer_profiler_incorrect_arg_type(profiler):
                        match=r"Only None, bool, str and subclasses of `BaseProfiler`"
                              r" are valid values for `Trainer`'s `profiler` parameter. *"):
         Trainer(profiler=profiler)
+
+
+@pytest.mark.parametrize(
+    ["limit_train_batches", "global_step", "num_training_batches", "current_epoch", "should_train"],
+    [pytest.param(0.2, 0, 0, 0, False), pytest.param(0.5, 10, 2, 4, True)],
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:1452]: Use tuples instead of pytest.param (unless there is a reason for it)

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -1444,3 +1445,65 @@ def test_trainer_profiler_incorrect_arg_type(profiler):
...
+        default_root_dir=tmpdir,
+        max_epochs=5,
+        limit_train_batches=limit_train_batches,
+    )
+    result = trainer.fit(model, train_loader)
+
+    after_state_dict = model.state_dict()
+
+    for key in before_state_dict.keys():
+        assert torch.all(torch.eq(before_state_dict[key], after_state_dict[key])) != should_train
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:1493]: I dont think we need to check the model state. The asserts at the bottom are enough

File: tests/trainer/test_trainer.py [hunk 7]
```diff
@@ -1444,3 +1445,65 @@ def test_trainer_profiler_incorrect_arg_type(profiler):
...
+
+    params_string = f"""`limit_train_batches={limit_train_batches}`, `dataset_len={dataset_len}`
+                        & `batch_size={batch_size}` as
+                        `num_training_batches={num_training_batches}`"""
+    if should_train:
+        error_string = f"should run with {params_string}"
+    else:
+        error_string = f"should not run with {params_string}"
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:1501]: I appreciate the effort you put into this error string but seems a bit overkill IMO

Comment (kaushikb11) [reply tests/trainer/test_trainer.py:1501]: What do you suggest instead? I don't see any harm in being descriptive with the error messages.

Comment (carmocca) [reply tests/trainer/test_trainer.py:1501]: Just not having it, pytest already reports the difference between "actual" and "expected"

But you can keep it if you want. No strong feelings about it 👍

File: tests/trainer/test_trainer.py (hunk 6 above)
Comment (kaushikb11) [reply tests/trainer/test_trainer.py]: Makes sense, thanks. Changes updated! @carmocca

**Review by carmocca — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/trainer/training_loop.py [hunk 8]
```diff
@@ -238,7 +232,7 @@ def on_train_batch_end(self, epoch_output, batch_end_outputs, batch, batch_idx,
         self.trainer.logger_connector.on_train_batch_end()
 
     def reset_train_val_dataloaders(self, model):
-        if not self.trainer.reload_dataloaders_every_epoch:
+        if self.trainer.train_dataloader is None or not self.trainer.reload_dataloaders_every_epoch:
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/training_loop.py:235]: was it failing somewhere if we don't update it?

**Review by rohitgr7 — APPROVED** [review]
lgtm!

File: pytorch_lightning/trainer/training_loop.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/training_loop.py]: going to do this in a followup PR?

Comment (kaushikb11) [reply pytorch_lightning/trainer/training_loop.py]: I will follow up after this [PR](https://github.com/PyTorchLightning/pytorch-lightning/pull/5726)

File: pytorch_lightning/trainer/training_loop.py (hunk 8 above)
Comment (kaushikb11) [reply pytorch_lightning/trainer/training_loop.py:235]: For the first epoch, when `reload_dataloaders_every_epoch` is True, hence the `dataloaders` are not reset before the training loop.

File: pytorch_lightning/trainer/training_loop.py [hunk 9]
```diff
@@ -203,7 +197,7 @@ def on_train_epoch_start(self, epoch):
         model = self.trainer.get_model()
 
         # reset train dataloader
-        if self.trainer.reload_dataloaders_every_epoch:
+        if epoch != 0 and self.trainer.reload_dataloaders_every_epoch:
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:200]:
```suggestion
        if epoch > 0 and self.trainer.reload_dataloaders_every_epoch:
```
it is there a case with negative value?

Comment (kaushikb11) [reply pytorch_lightning/trainer/training_loop.py:200]: I don't think there would an instance of negative values.
